### PR TITLE
Automatically select vLLM for matching GPUs

### PR DESCRIPTION
**Pull request by r0x0r**
This pull request introduces improved GPU-aware engine selection and enhanced robustness for ROCm CLI's model serving and SDK installation workflows. The main focus is on preferring the vLLM engine for supported AMD GPU families, automatically installing vLLM when appropriate, and providing better diagnostics and configurability for engine selection. Several tests are added to ensure correct behavior across edge cases.

Key changes include:

**Engine Selection and Auto-Installation Improvements:**

- Added logic to prefer the vLLM engine for serving models on supported ROCm GPU families (e.g., `gfx90a`, `gfx94X-dcgpu`). If the model recipe supports vLLM, it is selected automatically; otherwise, the recipe's preferred engine is used. This ensures optimal engine selection based on hardware capabilities. [[1]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R3255) [[2]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R3271-R3285) [[3]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R3408-R3414) [[4]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R11730-R11738)
- Introduced automatic installation of the vLLM engine immediately after SDK installation for compatible GPU families, including audit logging and user warnings if the install fails. [[1]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R1590-R1609) [[2]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R4669-R4750) [[3]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R4791)

**Model Reference and Engine Compatibility:**

- Enhanced `serve_model_ref_for_engine` to respect engine-specific model ID overrides, ensuring the correct model variant is served for each engine and avoiding incompatibility issues.

**Diagnostics and Robustness:**

- Improved error reporting for managed engine startup failures by including recent log output and retrying log reads to provide more actionable diagnostics. [[1]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R3566-R3588) [[2]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1L3610-R3683) [[3]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1L10849-R10999)
- Changed process termination logic to use a tree-based kill for more reliable cleanup of managed services.

**Testing and Code Quality:**

- Added comprehensive tests to verify engine selection logic, including GPU-based preferences, model/engine compatibility, and SDK install-time engine auto-selection. [[1]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1L17239-R17391) [[2]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R17410-R17488) [[3]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1L17276-R17508)
- Updated function signatures and usage throughout the codebase to support the new engine selection logic, and added missing fields to test data for completeness. [[1]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R17548) [[2]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R17557) [[3]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1R13771) [[4]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1L10546-R10695) [[5]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1L51-R53) [[6]](diffhunk://#diff-85923bafe93886cd11cf9fc2a11c96d715facfd36cea9a7018b79526241f40a1L26-R35)

These changes collectively improve ROCm CLI's ability to select and manage the best engine for the user's hardware, automate common setup steps, and provide clearer feedback when issues are encountered.

### CODE REVIEW

File: engines/vllm/src/lib.rs [hunk 1]
```diff
@@ -1233,6 +1373,102 @@ fn print_json<T: Serialize>(value: &T) -> Result<()> {
...
+    model_ref: &str,
+    timeout: Duration,
+    log_path: Option<&Path>,
+) -> Result<()> {
+    let start = SystemTime::now();
+    let endpoint = format!("http://{}:{}", host, port);
+    let poll_interval = Duration::from_millis(500);
+
+    loop {
+        if start.elapsed().unwrap_or(timeout) > timeout {
```
[the remark this answers is not in the document]
Comment (r0x0r) [reply engines/vllm/src/lib.rs]: Done in 7abd18a. `wait_for_vllm_ready` now tracks elapsed time with a monotonic `std::time::Instant` and the timeout check is simplified accordingly, so wall-clock adjustments can no longer corrupt the timeout.

File: engines/vllm/src/lib.rs [hunk 2]
```diff
@@ -497,13 +540,31 @@ struct ServeHttpRequest {
...
 }
 
 fn serve_http(request: ServeHttpRequest) -> Result<()> {
     let runtime = resolve_vllm_runtime(request.runtime_id.as_deref())?;
-    let mut child = spawn_vllm_server(&request, &runtime, None)?;
+    let mut child = spawn_vllm_server(&request, &runtime, request.log_path.as_deref())?;
     write_running_state(&request, &runtime, child.id())?;
+
+    // Wait for the server to become ready, with comprehensive error logging
+    if let Err(e) = wait_for_vllm_ready(
```
[the remark this answers is not in the document]
Comment (r0x0r) [reply engines/vllm/src/lib.rs:581]: Done in 7abd18a. The readiness loop now calls `child.try_wait()` at the top of each iteration and bails immediately (including the startup-log tail) when the vLLM process has already exited, so bad model refs / missing deps surface right away instead of waiting out the full timeout.

File: apps/rocm/src/main.rs [hunk 3]
```diff
@@ -3358,11 +3405,13 @@ fn serve(
...
     let paths = AppPaths::discover()?;
     let mut config = RocmCliConfig::load(&paths)?;
+    let host_gpu_summary = detect_host_gpu_summary(Some(&paths));
     let shared_recipe = resolve_model_recipe(&model)?;
     let serve_engine = select_serve_engine(
         engine.as_deref(),
         config.default_engine.as_deref(),
         shared_recipe.as_ref(),
+        Some(&host_gpu_summary),
     );
```
[the remark this answers is not in the document]
Comment (r0x0r) [reply apps/rocm/src/main.rs:3773]: Done in 7abd18a. `serve` now only calls `detect_host_gpu_summary()` when neither an explicit `--engine` nor a non-empty configured `default_engine` is present, i.e. only when GPU-based selection would actually consult it.

File: crates/rocm-core/src/lib.rs [hunk 4]
```diff
@@ -1130,6 +1212,40 @@ pub const fn default_engine_for_platform() -> &'static str {
...
+pub fn preferred_serve_engine_for_host_gpu_summary(
+    summary: &HostGpuSummary,
+) -> Option<&'static str> {
+    preferred_serve_engine_for_therock_family(
+        summary
+            .therock_family
+            .as_deref()
+            .or(summary.gfx_target.as_deref()),
+    )
+}
```
[the remark this answers is not in the document]
Comment (r0x0r) [reply crates/rocm-core/src/lib.rs:1245]: Done in 7abd18a. `preferred_serve_engine_for_host_gpu_summary` now returns `None` on native Windows (`cfg!(windows)`), matching the vLLM adapter which bails out there; Linux/WSL builds are unaffected. Added a regression test covering the platform gate.

**Comment by michaelroy-amd** [comment]
**Review — approve with nits.** Gates clean (clippy `-D warnings`, fmt, tests); no new files. GPU-family→vLLM selection, install auto-provisioning, and process-tree termination are well done and well tested.

- **[HIGH]** `wait_for_vllm_ready(..., Duration::from_mins(5))` in `serve_http` (engines/vllm/src/lib.rs) is new behavior — a valid-but-slow cold start (large weight download / first-decode compile) that exceeds 5 min now gets `terminate_process_tree`'d and marked `failed`. Regression for big models on exactly the datacenter GPUs this targets. Make the timeout configurable (env, like `ROCM_CLI_VLLM_ENFORCE_EAGER`) and/or reset on download/compile progress.
- **[MEDIUM]** Hardcoded `VLLM_ROCM_EXTRA_INDEX_URL = ".../rocm/0.23.0/rocm723"` pins both vLLM version and ROCm ABI; will silently drift from the resolved runtime. Derive the ROCm tag from the runtime manifest, or make it a documented const with an override.
- **[MEDIUM]** Reinstall path: confirm `uv_pip_install_base` actually passes `--reinstall`/`--force-reinstall`; otherwise `--reinstall` is a silent no-op when the wheel is present.
- **[LOW]** `tail_lines` can drop a complete line 0 when the seek lands on a boundary (>4 MB logs only). `maybe_auto_install_sdk_preferred_engine`'s `if engine != "vllm"` guard is dead today.

Nice handling of `ESRCH` / `/proc/<pid>/stat` parsing in the process-tree teardown.

**Comment by r0x0r** [comment]
Thanks for the review @michaelroy-amd. Addressed the nits in dc8f386:

- **[HIGH] 5-min readiness timeout regression.** `serve_http` no longer hardcodes the window — `wait_for_vllm_ready` now takes `vllm_ready_timeout()`, which is configurable via `ROCM_CLI_VLLM_READY_TIMEOUT_SECS` (positive integer seconds) and defaults to the previous 5 minutes. Slow-but-valid cold starts (large weight download / first-decode compile) can be given a larger budget instead of being `terminate_process_tree`'d.

- **[MEDIUM] Hardcoded `VLLM_ROCM_EXTRA_INDEX_URL`.** The const is now documented as pinning both the vLLM release and ROCm ABI tag, and `install_vllm_with_uv` resolves the index through `vllm_rocm_extra_index_url()`, which honors a `ROCM_CLI_VLLM_ROCM_INDEX_URL` override (non-empty, trimmed) so it can be pointed at a matching runtime without a rebuild.

- **[MEDIUM] Reinstall path no-op.** Confirmed `uv_pip_install_base` did not pass `--reinstall`, so `reinstall: true` was a silent no-op when the wheel was present. `install_response` now threads the flag into `install_vllm_with_uv`, which adds `--reinstall` when requested.

- **[LOW] `tail_lines` boundary drop.** The leading-fragment removal previously fired whenever the seek window was used, dropping a complete first line when the seek landed exactly on a newline boundary. It now probes the byte preceding `seek_pos` and only drops a genuinely partial fragment.

- **[LOW] dead `if engine != "vllm"` guard.** Removed; `maybe_auto_install_sdk_preferred_engine` now renders the preferred engine name generically.

Added regression tests for the timeout/index resolution helpers and for both `tail_lines` boundary cases (mid-line drop vs. on-boundary keep). Gates: `cargo clippy --workspace --all-targets -- -D warnings`, `cargo fmt --all --check`, touched-package tests, and `scripts/smoke_local.py` all pass.

**Review by juhovainio — APPROVED** [review]

**Comment by rominf** [comment]
Rebased onto latest `main` to pick up #49 (the new `--gpu` flag) and resolve the merge conflict.

The only non-trivial resolution was in `engines/vllm/src/lib.rs`, where the `ServeHttp` arm builds `ServeHttpRequest`: #49 added `gpu_indices: parse_gpu_indices_arg(gpu.as_deref())?` and this PR added the `log_path` computation — the rebase keeps both. Full workspace builds and `cargo clippy`/`cargo test` pass via the pre-push hooks.

Note: the force-push dismissed the prior approval since the head moved — a re-confirm would be appreciated, @juhovainio.
